client/controller/shared.js

```diff
@@ -15,13 +15,16 @@ import { isTranslatedIncompletely } from 'lib/i18n-utils/utils';
 
 export function makeLayoutMiddleware( LayoutComponent ) {
 	return ( context, next ) => {
-		const { store, primary, secondary } = context;
+		const { store, section, pathname, query, primary, secondary } = context;
 
 		// On server, only render LoggedOutLayout when logged-out.
 		if ( ! context.isServerSide || ! getCurrentUser( context.store.getState() ) ) {
 			context.layout = (
 				<LayoutComponent
 					store={ store }
+					currentSection={ section }
+					currentRoute={ pathname }
+					currentQuery={ query }
```

Comment:
Yes, if this PR is to be production-ready, and be able to completely replace the existing Redux implementation, it should somehow provide the `initial` and `previous` values, too.

I believe these values should be stored (and maybe already are) somewhere in the _page.js_ router or in `window.history` and only exposed by the React context API here.

It would also be nice to audit the existing `initial` and `previous` usages and learn more about why someone needs them at all. For example, why does anyone need to look at `previous` route? Quick look at existing usages suggests that some of them could be replaced by a simple `window.history.back()`.

Reply:
Agreed. Probably the best thing to do would be to understand the reasons for `initial` and `previous` to be there in the first place.
With that in mind, we can make a better decision on what's the best implementation.